File: packages/elixir/lib/kreuzberg/epub_metadata.ex
defmodule Kreuzberg.EpubMetadata do
  @moduledoc """
  EPUB metadata (Dublin Core extensions).

  Matches the Rust `EpubMetadata` struct.

  ## Fields

    * `:coverage` - Geographic/temporal coverage
    * `:dc_format` - Dublin Core format
    * `:relation` - Related resource
    * `:source` - Source resource
    * `:dc_type` - Dublin Core type
    * `:cover_image` - Cover image path
  """

  @type t :: %__MODULE__{
          coverage: String.t() | nil,
          dc_format: String.t() | nil,
          relation: String.t() | nil,
          source: String.t() | nil,
          dc_type: String.t() | nil,
          cover_image: String.t() | nil
        }

  defstruct coverage: nil,
            dc_format: nil,
            relation: nil,
            source: nil,
            dc_type: nil,
            cover_image: nil

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    %__MODULE__{
      coverage: data["coverage"],
      dc_format: data["dc_format"],
      relation: data["relation"],
      source: data["source"],
      dc_type: data["dc_type"],
      cover_image: data["cover_image"]
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = meta) do
    %{
      "coverage" => meta.coverage,
      "dc_format" => meta.dc_format,
      "relation" => meta.relation,
      "source" => meta.source,
      "dc_type" => meta.dc_type,
      "cover_image" => meta.cover_image
    }
  end
end

File: packages/php/src/Types/Metadata/EpubMetadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types\Metadata;

/**
 * EPUB metadata (Dublin Core extensions).
 *
 * Contains extended Dublin Core metadata fields specific to EPUB files.
 */
readonly class EpubMetadata
{
    public function __construct(
        public ?string $coverage = null,
        public ?string $dcFormat = null,
        public ?string $relation = null,
        public ?string $source = null,
        public ?string $dcType = null,
        public ?string $coverImage = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $rawCoverage = $data['coverage'] ?? null;
        $rawDcFormat = $data['dc_format'] ?? null;
        $rawRelation = $data['relation'] ?? null;
        $rawSource = $data['source'] ?? null;
        $rawDcType = $data['dc_type'] ?? null;
        $rawCoverImage = $data['cover_image'] ?? null;

        return new self(
            coverage: is_string($rawCoverage) ? $rawCoverage : null,
            dcFormat: is_string($rawDcFormat) ? $rawDcFormat : null,
            relation: is_string($rawRelation) ? $rawRelation : null,
            source: is_string($rawSource) ? $rawSource : null,
            dcType: is_string($rawDcType) ? $rawDcType : null,
            coverImage: is_string($rawCoverImage) ? $rawCoverImage : null,
        );
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [];

        if ($this->coverage !== null) {
            $result['coverage'] = $this->coverage;
        }

        if ($this->dcFormat !== null) {
            $result['dc_format'] = $this->dcFormat;
        }

        if ($this->relation !== null) {
            $result['relation'] = $this->relation;
        }

        if ($this->source !== null) {
            $result['source'] = $this->source;
        }

        if ($this->dcType !== null) {
            $result['dc_type'] = $this->dcType;
        }

        if ($this->coverImage !== null) {
            $result['cover_image'] = $this->coverImage;
        }

        return $result;
    }
}
